File: src/com/lapissea/vulkanimpl/simplevktypes/VkDescriptorSet.java
package com.lapissea.vulkanimpl.simplevktypes;

import com.lapissea.util.LogUtil;
import com.lapissea.vulkanimpl.Vk;
import com.lapissea.vulkanimpl.util.DevelopmentMemorySafety;
import com.lapissea.vulkanimpl.util.VkDestroyable;

import java.nio.LongBuffer;

import static org.lwjgl.system.MemoryUtil.*;

public class VkDescriptorSet extends ExtendableLong implements VkDestroyable{
	
	private LongBuffer pointer;
	
	public VkDescriptorSet(LongBuffer val){
		super(val.get(0));
		pointer=val;
	}
	
	public LongBuffer pointer(){
		return pointer;
	}
	
	@Override
	public void destroy(){
		memFree(pointer);
		pointer=null;
	}
	
	private final StackTraceElement[] _ctorStack=Vk.DEVELOPMENT?DevelopmentMemorySafety.make():null;
	
	@Override
	protected void finalize() throws Throwable{
		if(Vk.DEVELOPMENT) DevelopmentMemorySafety.check(this, pointer);
	}
}

File: src/com/lapissea/vulkanimpl/util/types/VkDescriptorSet.kt
package com.lapissea.vulkanimpl.util.types

import com.lapissea.vulkanimpl.VkGpu
import com.lapissea.vulkanimpl.util.VkBufferedHandle
import com.lapissea.vulkanimpl.util.VkDestroyable
import com.lapissea.vulkanimpl.util.VkGpuCtx
import com.lapissea.vulkanimpl.util.format.VkDescriptor
import org.lwjgl.system.MemoryStack
import org.lwjgl.system.MemoryUtil.*
import org.lwjgl.vulkan.VK10.*
import org.lwjgl.vulkan.VkCopyDescriptorSet
import org.lwjgl.vulkan.VkWriteDescriptorSet
import java.nio.LongBuffer

class VkDescriptorSet(private val gpu: VkGpu, private val buff:LongBuffer): VkDestroyable, VkGpuCtx, VkBufferedHandle {
	
	override fun getBuff()=buff
	override fun getGpu()=gpu
	
	override fun destroy() {
		buff.put(0,0)
		memFree(buff)
	}
	
	
	fun update(descriptors: Array<VkDescriptor>) {
		MemoryStack.stackPush().use {
			return update(VkDescriptor.createWriteDescriptorSet(descriptors, this, it))
		}
	}
	fun update(descriptorWrites: VkWriteDescriptorSet.Buffer) {
		update(descriptorWrites, null)
	}
	
	fun update(descriptorWrites: VkWriteDescriptorSet.Buffer, descriptorCopies: VkCopyDescriptorSet.Buffer?) {
		vkUpdateDescriptorSets(device, descriptorWrites, descriptorCopies)
	}
}
